src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/ChainedMethodCallWrapCheck.java

```diff
@@ -0,0 +1,382 @@
+        DetailAST childAst = methodCallAst.getFirstChild();
+        while (TokenUtil.isOfType(childAst, METHOD_CHAIN_CONTINUATION_NODES)) {
+            childAst = getMethodCallDescendantAst(childAst);
+        }
+        return TokenUtil.areOnSameLine(childAst, methodCallAst);
+    }
+
+    /**
+     * Check chained method calls spanning over a single line.
+     *
+     * @param methodCallAst ast node of type {@link TokenTypes#METHOD_CALL}
+     */
+    private void checkMethodChainingSingleLine(DetailAST methodCallAst) {
+        DetailAST childAst = methodCallAst.getFirstChild();
+        int methodCount = 1;
+        while (TokenUtil.isOfType(childAst, METHOD_CHAIN_CONTINUATION_NODES)) {
+            if (childAst.getType() == TokenTypes.METHOD_CALL) {
+                methodCount++;
+            }
+            childAst = getMethodCallDescendantAst(childAst);
+        }
+        if (methodCount > maxSingleLineCalls) {
+            log(childAst, MSG_TOO_MANY_SINGLE_LINE, methodCount, maxSingleLineCalls);
+        }
+    }
+
+    /**
+     * Check chained method calls spanning over multiple lines.
+     *
+     * @param methodCallAst ast node of type {@link TokenTypes#METHOD_CALL}
+     */
+    private void checkMethodChainingMultiLine(DetailAST methodCallAst) {
+        DetailAST childAst = methodCallAst.getFirstChild();
+        DetailAST currentLineOuterMostMethodAst = methodCallAst;
+        int methodCount = 1;
+        while (TokenUtil.isOfType(childAst, METHOD_CHAIN_CONTINUATION_NODES)) {
+            if (childAst.getType() == TokenTypes.METHOD_CALL) {
+                final int methodCallLineNo = childAst.getLineNo();
+                if (currentLineOuterMostMethodAst.getLineNo() == methodCallLineNo) {
+                    methodCount++;
+                    if (methodCount > maxCallsPerLine) {
+                        log(currentLineOuterMostMethodAst, MSG_WRAP_METHOD_CALL);
+                        currentLineOuterMostMethodAst = childAst;
+                    }
+                }
+                else {
+                    currentLineOuterMostMethodAst = childAst;
+                    methodCount = 1;
+                }
+            }
+            childAst = getMethodCallDescendantAst(childAst);
+        }
+    }
+
+    /**
+     * Get the required descendant ast of method call ast.
+     *
+     * @param ast ast
+     * @return required descendant ast of method call ast.
+     */
+    private static DetailAST getMethodCallDescendantAst(DetailAST ast) {
+        DetailAST childAst = ast.getFirstChild();
+        while (childAst.getType() == TokenTypes.LPAREN) {
+            childAst = childAst.getNextSibling();
+            if (!TokenUtil.isOfType(childAst, METHOD_CHAIN_CONTINUATION_NODES)) {
+                childAst = getMethodCallDescendantAfterLparen(childAst);
+            }
+        }
+        return childAst;
+    }
+
+    /**
+     * Get method call descendant after encountering {@link TokenTypes#LPAREN}
```

Comment:
It would be good to describe why this requires special consideration for future maintianers.

Reply:
Done.

Reply:
Resolved.